### DISCUSSION TITLE
While multiclass, ValueError: cannot reshape array of size 10 into shape (2,2)@wangqingl If you have 10 classes you need to write separate visualizer to visualize all the classes. Since, you removed `pred * 255` your values the `output.png` will be between 0 and 9 - hence you see dark (or black) output. You need to write the visualizer such that it will assign different color values for each predicted output class.

### DISCUSSION THREAD

**Discussion by StorywithLove** [wgcban/ChangeFormer#94] [Announcements]
When I made modifications according to the above method, the following error occurred during training. How to solve it? @wangqingl If you have 10 classes you need to write separate visualizer to visualize all the classes. Since, you removed `pred * 255` your values the `output.png` will be between 0 and 9 - hence you see dark (or black) output. You need to write the visualizer such that it will assign different color values for each predicted output class.

_Originally posted by @wgcban in https://github.com/wgcban/ChangeFormer/issues/88#issuecomment-1856013788_

**Comment by StorywithLove** [comment]
So far it's been found that it's due to hyperparameters not being passed into the function correctly, and the solution is as follows:

![image](https://github.com/wgcban/ChangeFormer/assets/48684315/2c03d64c-90ad-4ec6-8642-7d14ffc3856f)

**Comment by StorywithLove** [comment]
Summary of the changes needed to modify binary change detection to multi-class change detection:

(1) run_ChangeFormer_cj.sh
    n_class=8 and make it a hyperparameter to python main.py

(2) models/networks.py
    net = ChangeFormerV6(embed_dim=args.embed_dim, output_nc=args.n_class)

 (3) models/basic_model.py  
    Comment out: pred_vis = pred * 255, i.e., modifications to visualisation processing

(4) models/trainer.py  
    Modify: ConfuseMatrixMeter(n_class=self.n_class)

**Comment by wgcban** [comment]
@Programming-Music Thank you for summarizing all these steps! It will be really helpful for those who wish to use ChangeFormer for multi-class change detection.

**Comment by StorywithLove** [comment]
When using ChangeFormer for multi-class change detection, I made the above adjustments to the model; however, the model is training with G_loss decreasing, but the metrics are all zero. I've tried turning down the learning rate, but I still can't fix the problem. Is there any potentially effective solutions?@wgcban

![image](https://github.com/wgcban/ChangeFormer/assets/48684315/3d8930cc-0edc-476a-8319-50a2d3560a7d)

**Reply by StorywithLove** [reply]
After some debugging and modifications, the model works perfectly on the multivariate change detection virtual dataset (randomly assigning labels to 0-N)!!!

Let's continue where we left off.(https://github.com/wgcban/ChangeFormer/discussions/94#discussioncomment-8321604)：

"(3) models/basic_model.py
Comment out: pred_vis = pred * 255, i.e., modifications to visualisation processing"
Perform the same for the _visualize_pred() function in models/tariner.py

(5) data_config.py
Comment out self.label_transform = "norm", i.e., do not execute "if self.label_transform == 'norm'.
            label = label // 255"

(6) models/train.py
Adjust the BCD output of the _collect_running_batch_states function as follows:
```
  # vis = np.concatenate([vis_input, vis_input2, vis_pred, vis_gt], axis=0)
  # vis = np.clip(vis, a_min=0.0, a_max=1.0)  
  vis = np.concatenate([(vis_input*255).astype(np.uint8), (vis_input2*255).astype(np.uint8), vis_pred.astype(np.uint8), vis_gt.astype(np.uint8)], axis=0)

  # plt.imsave(file_name, vis)
  Image.fromarray(vis).save(file_name)
```
![image](https://github.com/wgcban/ChangeFormer/assets/48684315/dbcc26d6-0189-45ec-8e4e-207dc5a9b7b2)
![c5944abc17f9067b6fc5f8e1b7da731e](https://github.com/wgcban/ChangeFormer/assets/48684315/a1dc0f2f-b205-42eb-9390-a59d85bcbf4b)

Printing out the output of the model at each epoch while adjusting, it helped me finish solving the problem. Thanks again to the author for the original code, outstanding work!

**Reply by wgcban** [reply]
Great! Glad you be able to modify it for multi class problem.

**Comment by hhu008994-pixel** [comment]
Hi, the pretrained_changeformer.pt Dropbox link seems expired. Could you please re-upload it?
